File: src/main/java/maifetch/AsciiLogo.java
package maifetch;

import javax.imageio.ImageIO;
import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.RenderingHints;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public final class AsciiLogo {
    private static final char[] RAMP = "@%#*+=-:. ".toCharArray();

    private AsciiLogo() {
    }

    public static List<String> fromUrl(String imageUrl, int size) throws IOException {
        BufferedImage image = ImageIO.read(new URL(imageUrl));
        if (image == null) {
            throw new IOException("could not decode profile icon");
        }
        int width = Math.max(1, size * 2);
        int height = Math.max(1, size);
        Image scaled = image.getScaledInstance(width, height, Image.SCALE_SMOOTH);
        BufferedImage target = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D graphics = target.createGraphics();
        graphics.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BILINEAR);
        graphics.drawImage(scaled, 0, 0, null);
        graphics.dispose();

        List<String> rows = new ArrayList<String>();
        for (int y = 0; y < height; y++) {
            StringBuilder row = new StringBuilder();
            for (int x = 0; x < width; x++) {
                int rgb = target.getRGB(x, y);
                int red = (rgb >> 16) & 0xFF;
                int green = (rgb >> 8) & 0xFF;
                int blue = rgb & 0xFF;
                int luminance = (red * 299 + green * 587 + blue * 114) / 1000;
                int index = luminance * (RAMP.length - 1) / 255;
                char ch = RAMP[index];
                row.append(Ansi.foreground(String.valueOf(ch == ' ' ? '#' : ch), red, green, blue));
            }
            rows.add(row.toString());
        }
        return rows;
    }
}

File: src/main/kotlin/maifetch/AsciiLogo.kt
package maifetch

import java.awt.Image
import java.awt.RenderingHints
import java.awt.image.BufferedImage
import java.net.URL
import javax.imageio.ImageIO

object AsciiLogo {
    private val ramp = "@%#*+=-:. ".toCharArray()

    fun fromUrl(imageUrl: String, size: Int): List<String> {
        val image = ImageIO.read(URL(imageUrl)) ?: throw java.io.IOException("could not decode profile icon")
        val width = maxOf(1, size * 2)
        val height = maxOf(1, size)
        val scaled = image.getScaledInstance(width, height, Image.SCALE_SMOOTH)
        val target = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        val graphics = target.createGraphics()
        graphics.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BILINEAR)
        graphics.drawImage(scaled, 0, 0, null)
        graphics.dispose()

        return (0 until height).map { y ->
            buildString {
                for (x in 0 until width) {
                    val rgb = target.getRGB(x, y)
                    val red = (rgb shr 16) and 0xFF
                    val green = (rgb shr 8) and 0xFF
                    val blue = rgb and 0xFF
                    val luminance = (red * 299 + green * 587 + blue * 114) / 1000
                    val index = luminance * (ramp.size - 1) / 255
                    val ch = if (ramp[index] == ' ') '#' else ramp[index]
                    append(Ansi.foreground(ch.toString(), red, green, blue))
                }
            }
        }
    }
}
